larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,462 @@
+        if(isAvailableU)
+            pfoParameters.m_clusterList.push_back(m_newClustersUMap.at(iCluster));
+        if(isAvailableV)
+            pfoParameters.m_clusterList.push_back(m_newClustersVMap.at(iCluster));
+        if(isAvailableW)
+            pfoParameters.m_clusterList.push_back(m_newClustersWMap.at(iCluster));
+
+        pfoParameters.m_clusterList.push_back(pNewThreeDCluster);
+        
+        pfoParameters.m_particleId = pPfoToRebuild->GetParticleId();
+        pfoParameters.m_charge = PdgTable::GetParticleCharge(pfoParameters.m_particleId.Get());
+        pfoParameters.m_mass = PdgTable::GetParticleMass(pfoParameters.m_particleId.Get());
+        pfoParameters.m_energy = 0.f;
+        pfoParameters.m_momentum = CartesianVector(0.f, 0.f, 0.f);
+        
+        const ParticleFlowObject *pNewPfo(nullptr);
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ParticleFlowObject::Create(*this, pfoParameters, pNewPfo));
+        
+        iCluster++;
+    }
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Pfo>(*this, newPfoListName, m_PfosForReclusteringListName));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::RebuildPfo(const Pfo *pPfoToRebuild, ClusterList &newClustersList)
+{
+    m_newClustersUMap.clear();
+    m_newClustersVMap.clear();
+    m_newClustersWMap.clear();
+
+    //For each of the new 3D clusters, build a new 2D cluster in each view, and a new Pfo
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->BuildNewTwoDClusters(pPfoToRebuild, newClustersList));
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->BuildNewPfos(pPfoToRebuild, newClustersList));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool ThreeDReclusteringAlgorithm::PassesCutsForReclustering(const pandora::ParticleFlowObject *const pShowerPfo)
+{
+    if (!LArPfoHelper::IsShower(pShowerPfo)) return false;
+    ClusterList clusterList3D;
+    LArPfoHelper::GetThreeDClusterList(pShowerPfo, clusterList3D);
+
+    if (clusterList3D.empty()) return false;
+    CaloHitList caloHitList3D;
+    clusterList3D.front()->GetOrderedCaloHitList().FillCaloHitList(caloHitList3D);
+
+    //Quality cuts
+    if (caloHitList3D.size() < 2) return false;
+
+    //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+    const ClusterList *pShowerClusters(nullptr);
+    PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+    if(!pShowerClusters) return false;
+
+    if(pShowerClusters->end() == std::find(pShowerClusters->begin(), pShowerClusters->end(), clusterList3D.front())) return false;
```

Comment:
Yeah, this is redundant. I think I'll remove this one and leave the one in "run" as it's a sanity check (as you point out, shouldn't happen at all) rather than a reclustering cut.